homeassistant/helpers/device_utils.py

```diff
@@ -0,0 +1,82 @@
+"""Provides useful helpers for dealing with devices."""
+
+from homeassistant.core import HomeAssistant, callback
+
+from . import device_registry as dr, entity_registry as er
+
+
+@callback
+def async_entity_id_to_device_id(
+    hass: HomeAssistant,
+    entity_id: str,
+) -> str | None:
+    """Resolve the device id to the entity id."""
+
+    ent_reg = er.async_get(hass)
+    if (entity := ent_reg.async_get(entity_id)) is None:
+        return None
+    return entity.device_id
+
+
+@callback
+def async_device_info_to_link(
+    hass: HomeAssistant,
+    device_id: str | None = None,
+    entity_id: str | None = None,
+) -> dr.DeviceInfo | None:
+    """DeviceInfo with the information to link a device in a config entry in the Link category.
+
+    It can directly receive a device ID or an entity ID to be extracted from the device.
+    """
+
+    dev_reg = dr.async_get(hass)
+
+    if device_id is None and entity_id is not None:
+        device_id = async_entity_id_to_device_id(hass, entity_id=entity_id)
```

Comment:
Same as for the other helper. Don't overload the function. Just have it work on one of the cases.